exercises/hardware_access_control/README.md

```diff
@@ -0,0 +1,160 @@
+
+```
+Led Walk Dynamic:           LED 3 Handle: 0x1087d0 (v:1 0x1087d0-0x1087e0 l:0x10 o:0xc p: G RWcgm- -- ---)
+Led Walk Dynamic: Destroyed LED 3 Handle: 0x1087d0 (v:1 0x1087d0-0x1087e0 l:0x10 o:0xc p: G RWcgm- -- ---)
+Led Walk Dynamic:       New LED 3 Handle: 0x108878 (v:1 0x108878-0x108888 l:0x10 o:0xc p: G RWcgm- -- ---)
+```
+
+These come from some superfluous lines in [`led_walk_dynamic.cc`][], that releases ownership of user LED 3 only to later reacquire ownership.
+You can comment out the line that reacquires the LED:
+
+```cpp
+	leds[3] = aquire_led(3).value();
+```
+
+When run `led_walk_dynamic` will now fail to toggle user LED 3 because it has relinquished ownership of the LED.
+
+```
+hardware_access_control/part_2/led_walk_dynamic.cc:34 Assertion failure in start_walking
+Failed to toggle an LED
+```
+
+
+## Part 3
+
+This is great and all, but how do we stop a compartment bypassing `gpio_access` and using `MMIO_CAPABILITY` directly?
+In other words, how do we ensure that only `gpio_access` has access to the GPIOs?
+
+Luckily the linker has all the information needed to check which compartments can access the GPIO MMIOs.
+It outputs this information in a JSON report with the rest of the build artefacts.
+To automate checking this report, we can use `cheriot-audit` which should already be in your path.
+
+`cheriot-audit` allows you to query the JSON report and assert certain rules are followed.
+You do this with a language called [Rego][], but don't worry you won't have to learn it for this exercise.
+There are some pre-written rules in the [`gpio_access.rego`] module.
+Let's first look at `only_gpio_access_has_access`.
+It uses `mmio_allow_list` from `cheriot-audit`'s included compartment package to check only the `gpio_access` compartment has access to the GPIO MMIOs.
+If we run this on the part 2 firmware image's JSON report, it will return true.
+However, when run against the part 1 firmware image's report it will return false, because the `blinky_raw` and `led_walk_raw` are not in the allow list.
+
+[Rego]: https://www.openpolicyagent.org/docs/latest/policy-language/
+[`gpio_access.rego`]: ../../exercises/hardware_access_control/part_3/gpio_access.rego
+
+```sh
+# A convenience function
+check_gpio_access() {
+    cheriot-audit \
+        --board cheriot-rtos/sdk/boards/sonata-prerelease.json \
+        --module exercises/hardware_access_control/part_3/gpio_access.rego \
+        --query "data.gpio_access.$1" \
+        --firmware-report "$2"
+}
+# This should return true
+check_gpio_access only_gpio_access_has_access \
+    build/cheriot/cheriot/release/hardware_access_part_2.json
+# This should return false
+check_gpio_access only_gpio_access_has_access \
+    build/cheriot/cheriot/release/hardware_access_part_1.json
+```
+
+There's a second rule, `whitelisted_compartments_only`, which adds an addition condition that only `led_walk_dynamic` and `blinky_dynamic` can use `gpio_access`.
```

Comment:
> which adds an addition condition

Typo: Which adds an **additional** condition.

Reply:
Addressed in: https://github.com/lowRISC/sonata-software/pull/42